### PR TITLE
bugfix for annotation updater.

### LINKED ISSUE #488: sonobuoy status doesn't report the tests are completed when the logs do.
**What steps did you take and what happened:**

I have automation running sonobuoy that has been using the logs to determine when the tests are finished and the results are available.  When I changed it to use `sonobuoy status`, I've encountered a case where `sonobuoy status` thinks the tests are still running when the logs indicate they're finished:

```
$ ./sonobuoy status
PLUGIN		STATUS		COUNT
e2e		running		1
systemd_logs	complete	5

Sonobuoy is still running. Runs can take up to 60 minutes.
```

```
$ kubectl logs -n heptio-sonobuoy sonobuoy -f
time="2018-07-17T17:48:10Z" level=info msg="Scanning plugins in ./plugins.d (pwd: /)"
time="2018-07-17T17:48:10Z" level=info msg="Scanning plugins in /etc/sonobuoy/plugins.d (pwd: /)"
time="2018-07-17T17:48:10Z" level=info msg="Directory (/etc/sonobuoy/plugins.d) does not exist"
time="2018-07-17T17:48:10Z" level=info msg="Scanning plugins in ~/sonobuoy/plugins.d (pwd: /)"
time="2018-07-17T17:48:10Z" level=info msg="Directory (~/sonobuoy/plugins.d) does not exist"
time="2018-07-17T17:48:10Z" level=info msg="Filtering namespaces based on the following regex:.*|heptio-sonobuoy"
time="2018-07-17T17:48:10Z" level=info msg="Namespace default Matched=true"
time="2018-07-17T17:48:10Z" level=info msg="Namespace e2e-tests-nettest-5qgxw Matched=true"
time="2018-07-17T17:48:10Z" level=info msg="Namespace heptio-sonobuoy Matched=true"
time="2018-07-17T17:48:10Z" level=info msg="Namespace kube-public Matched=true"
time="2018-07-17T17:48:10Z" level=info msg="Namespace kube-system Matched=true"
time="2018-07-17T17:48:10Z" level=info msg="Starting server Expected Results: [{ e2e} {fab10-compute-1 systemd_logs} {fab10-compute-2 systemd_logs} {fab10-compute-3 systemd_logs} {fab10-compute-4 systemd_logs} {fab10-compute-5 systemd_logs}]"
time="2018-07-17T17:48:10Z" level=info msg="starting aggregation server" address=0.0.0.0 port=8080
time="2018-07-17T17:48:10Z" level=info msg="Running plugin" plugin=e2e
time="2018-07-17T17:48:10Z" level=info msg="Running plugin" plugin=systemd-logs
time="2018-07-17T17:48:13Z" level=info msg="received aggregator request" client_cert=systemd-logs node=fab10-compute-1 plugin_name=systemd_logs
time="2018-07-17T17:48:13Z" level=info msg="received aggregator request" client_cert=systemd-logs node=fab10-compute-3 plugin_name=systemd_logs
time="2018-07-17T17:48:13Z" level=info msg="received aggregator request" client_cert=systemd-logs node=fab10-compute-2 plugin_name=systemd_logs
time="2018-07-17T17:48:13Z" level=info msg="received aggregator request" client_cert=systemd-logs node=fab10-compute-5 plugin_name=systemd_logs
time="2018-07-17T17:48:13Z" level=info msg="received aggregator request" client_cert=systemd-logs node=fab10-compute-4 plugin_name=systemd_logs
time="2018-07-17T17:48:30Z" level=info msg="received aggregator request" client_cert=e2e plugin_name=e2e
time="2018-07-17T17:48:30Z" level=info msg="Running non-ns query"
time="2018-07-17T17:48:30Z" level=error msg="error querying CustomResourceDefinitions: the server could not find the requested resource"
time="2018-07-17T17:48:30Z" level=info msg="Collecting Node Configuration and Health..."
time="2018-07-17T17:48:30Z" level=info msg="Creating host results for fab10-compute-1 under /tmp/sonobuoy/d6b2dd57-773c-4823-8e3c-69fc80a4b137/hosts/fab10-compute-1\n"
time="2018-07-17T17:48:30Z" level=warning msg="Could not get configz endpoint for node fab10-compute-1: the server could not find the requested resource"
time="2018-07-17T17:48:30Z" level=warning msg="Could not get healthz endpoint for node fab10-compute-1: the server could not find the requested resource"
time="2018-07-17T17:48:30Z" level=info msg="Creating host results for fab10-compute-2 under /tmp/sonobuoy/d6b2dd57-773c-4823-8e3c-69fc80a4b137/hosts/fab10-compute-2\n"
time="2018-07-17T17:48:30Z" level=warning msg="Could not get configz endpoint for node fab10-compute-2: the server could not find the requested resource"
time="2018-07-17T17:48:30Z" level=warning msg="Could not get healthz endpoint for node fab10-compute-2: the server could not find the requested resource"
time="2018-07-17T17:48:30Z" level=info msg="Creating host results for fab10-compute-3 under /tmp/sonobuoy/d6b2dd57-773c-4823-8e3c-69fc80a4b137/hosts/fab10-compute-3\n"
time="2018-07-17T17:48:30Z" level=warning msg="Could not get configz endpoint for node fab10-compute-3: the server could not find the requested resource"
time="2018-07-17T17:48:30Z" level=warning msg="Could not get healthz endpoint for node fab10-compute-3: the server could not find the requested resource"
time="2018-07-17T17:48:30Z" level=info msg="Creating host results for fab10-compute-4 under /tmp/sonobuoy/d6b2dd57-773c-4823-8e3c-69fc80a4b137/hosts/fab10-compute-4\n"
time="2018-07-17T17:48:30Z" level=warning msg="Could not get configz endpoint for node fab10-compute-4: the server could not find the requested resource"
time="2018-07-17T17:48:31Z" level=warning msg="Could not get healthz endpoint for node fab10-compute-4: the server could not find the requested resource"
time="2018-07-17T17:48:31Z" level=info msg="Creating host results for fab10-compute-5 under /tmp/sonobuoy/d6b2dd57-773c-4823-8e3c-69fc80a4b137/hosts/fab10-compute-5\n"
time="2018-07-17T17:48:31Z" level=warning msg="Could not get configz endpoint for node fab10-compute-5: the server could not find the requested resource"
time="2018-07-17T17:48:31Z" level=warning msg="Could not get healthz endpoint for node fab10-compute-5: the server could not find the requested resource"
time="2018-07-17T17:48:31Z" level=error msg="error querying ThirdPartyResources: don't know how to handle non-namespaced resource ThirdPartyResources"
time="2018-07-17T17:48:31Z" level=info msg="Running ns query (default)"
time="2018-07-17T17:48:33Z" level=info msg="Collecting Pod Logs..."
time="2018-07-17T17:48:33Z" level=error msg="error querying PodPresets: the server could not find the requested resource (get podpresets.settings.k8s.io)"
time="2018-07-17T17:48:35Z" level=info msg="Running ns query (e2e-tests-nettest-5qgxw)"
time="2018-07-17T17:48:36Z" level=info msg="Collecting Pod Logs..."
time="2018-07-17T17:48:37Z" level=error msg="error querying PodPresets: the server could not find the requested resource (get podpresets.settings.k8s.io)"
time="2018-07-17T17:48:38Z" level=info msg="Running ns query (heptio-sonobuoy)"
time="2018-07-17T17:48:39Z" level=info msg="Collecting Pod Logs..."
time="2018-07-17T17:48:42Z" level=error msg="error querying PodPresets: the server could not find the requested resource (get podpresets.settings.k8s.io)"
time="2018-07-17T17:48:44Z" level=info msg="Running ns query (kube-public)"
time="2018-07-17T17:48:45Z" level=info msg="Collecting Pod Logs..."
time="2018-07-17T17:48:45Z" level=error msg="error querying PodPresets: the server could not find the requested resource (get podpresets.settings.k8s.io)"
time="2018-07-17T17:48:46Z" level=info msg="Running ns query (kube-system)"
time="2018-07-17T17:48:48Z" level=info msg="Collecting Pod Logs..."
time="2018-07-17T17:49:11Z" level=error msg="error querying PodPresets: the server could not find the requested resource (get podpresets.settings.k8s.io)"
time="2018-07-17T17:49:13Z" level=info msg="Results available at /tmp/sonobuoy/201807171748_sonobuoy_d6b2dd57-773c-4823-8e3c-69fc80a4b137.tar.gz"
time="2018-07-17T17:49:13Z" level=info msg="no-exit was specified, sonobuoy is now blocking"
```

**What did you expect to happen:**

`sonobuoy status` to show the tests have completed.

**Environment:**

- Sonobuoy version: 0.11.3
- Kubernetes version: (use `kubectl version`): 1.10.5
- Kubernetes installer & version: kubespray
- Cloud provider or hardware configuration: bare metal
- OS (e.g. from `/etc/os-release`): Ubuntu 16.04
- Sonobuoy tarball (which contains * below)
[201807171748_sonobuoy_d6b2dd57-773c-4823-8e3c-69fc80a4b137.tar.gz](https://github.com/heptio/sonobuoy/files/2202847/201807171748_sonobuoy_d6b2dd57-773c-4823-8e3c-69fc80a4b137.tar.gz)

### PR DESCRIPTION
**Pull request by timothysc**
**What this PR does / why we need it**:
Fixes annotation updater issue(s).

**Which issue(s) this PR fixes**
- Fixes #488

**Special notes for your reviewer**:
I still need to run some tests on this to verify.   

**Release note**:
```
Will try to always update annotations on every exit condition. 
```

/assign @liztio

### CODE REVIEW

File: pkg/plugin/aggregation/run.go [hunk 1]
```diff
@@ -113,21 +116,31 @@ func Run(client kubernetes.Interface, plugins []plugin.Interface, cfg plugin.Agg
...
-		for range ticker.C {
-			updater.ReceiveAll(aggr.Results)
-			if err := updater.Annotate(); err != nil {
-				logrus.WithError(err).Info("couldn't annotate sonobuoy pod")
-			}
-			if aggr.isComplete() {
-				return
-			}
+	wait.JitterUntil(func() {
+		pluginsdone := aggr.isComplete()
```
Comment (timothysc) [inline pkg/plugin/aggregation/run.go]: I could probably make this a var outside and used on final check in defer.

File: pkg/plugin/aggregation/run.go [hunk 2]
```diff
@@ -113,21 +116,31 @@ func Run(client kubernetes.Interface, plugins []plugin.Interface, cfg plugin.Agg
...
+		updater.ReceiveAll(aggr.Results)
+		if err := updater.Annotate(); err != nil {
+			logrus.WithError(err).Info("couldn't annotate sonobuoy pod")
 		}
-	}()
+		if pluginsdone {
+			logrus.Info("All plugins have completed, status has been updated")
+			cancel()
+		}
+	}, annotationUpdateFreq, 1.2, true, ctx.Done())
```
Comment (liztio) [inline pkg/plugin/aggregation/run.go]: can you const these magic numbers?

File: pkg/plugin/aggregation/run.go [hunk 3]
```diff
@@ -113,21 +116,31 @@ func Run(client kubernetes.Interface, plugins []plugin.Interface, cfg plugin.Agg
...
-	ticker := time.NewTicker(annotationUpdateFreq)
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer func() {
+		logrus.Info("Last update to annotations on exit")
+		// This is the async exit cleanup function.
+		// 1. Stop the annotation updater
+		cancel()
+		// 2. Try one last time to get an update out on exit
+		updater.ReceiveAll(aggr.Results)
+		if err := updater.Annotate(); err != nil {
```
Comment (liztio) [inline pkg/plugin/aggregation/run.go]: could ReceiveAll() and Annotate() be combined? looks like they're duplicated.

File: pkg/plugin/aggregation/run.go [hunk 4]
```diff
@@ -113,21 +116,31 @@ func Run(client kubernetes.Interface, plugins []plugin.Interface, cfg plugin.Agg
...
-				return
-			}
+	wait.JitterUntil(func() {
+		pluginsdone := aggr.isComplete()
+		updater.ReceiveAll(aggr.Results)
+		if err := updater.Annotate(); err != nil {
+			logrus.WithError(err).Info("couldn't annotate sonobuoy pod")
 		}
-	}()
+		if pluginsdone {
```
Comment (liztio) [inline pkg/plugin/aggregation/run.go]: any particular reason we emit this on the JitterUntil case but not on the termination guard case?

**Pull request merged**